plugins/javaembed/HpccClassLoader.java

```diff
@@ -74,25 +87,55 @@ public synchronized Class<?> findClass(String className) throws ClassNotFoundExc
     {
         String luName = className.replace(".","/");
         Class<?> result = classes.get(luName);
-        if (result == null)
+        if (result != null)
+            return result;
+        if (trace)
```

Comment:
Although it may not be strictly necessary, I think it would be good to try to find the class in the parent class loader at this point by calling `super.findClass()`.

The standard practice for Java class loaders is to first look in the local cache of resolved classes and return it if found.  Next they ask the parent class loaders to to find the class.  If that fails then they use their own logic for finding the class.  

I think that this code will actually do that but it's kind of roundabout.  When it gets into the loop below it will call the first libraryLoader and that loader will first ask it's parent loader.  Since all of the library loaders should have the same parent class loader, this probably works.  But I think it would be better to be explicitly follow the standard practice and call the parent explicitly first.

Reply:
Yes, I think it should. The old code used to have that call in, though it would only have been hit  if there were no pathloaders active I suspect.

Reply:
I'm not sure why standard parctice would be "cache then parent then own logic" rather than "cached then own then parent", given I want any classes that can be found by my own logic to override those that might be found in the default loader.

I also wonder whether I should be caching nohits as well as hits.

Reply:
This SO answer describes the reason why the standard is 'parent then own': https://stackoverflow.com/a/51713294.  There are cases where that can be reversed but you have to be very careful.  I accidentally did the wrong order for a class loader for the KEL Intellij plugin and it definitely didn't work.

But in this case since the 'host' is not a Java application there is most likely no problem with 'own then parent' as long as the workunit jar doesn't have any classes defined in the bootstrap loader (i.e. the built in Java classes).  But 'parent then own' shouldn't be a problem either because the parent should really only have the built-in and HPCCUtil classes which shouldn't be in the work unit jars.

Most class loaders don't cache nohits maybe because most code dies if a class can't be loaded.  But I can't see any reason to not cache them for this use case.  As long as the Java code doesn't have problems, it shouldn't matter but if a bad jar is deployed it might reduce load if it's getting hit frequently.

Reply:
I read the SO you cited (and a few others!) and have gone from "I'm not sure I really understand how class loading in Java works" to "I am SURE I don't understand how class loading in java works".